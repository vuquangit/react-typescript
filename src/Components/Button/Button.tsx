import React, { FC, forwardRef } from 'react'
import { useSelector } from 'react-redux'
import { RootState } from '@/stores/rootReducer'

import { AppButton } from './Button.styled'
import { IButtonProps } from './Button.types'
const Button: FC<IButtonProps> = forwardRef((props, ref) => {
  const themeMode = useSelector((state: RootState) => state.theme.themeMode)

  return (
    <AppButton
      data-testid="app-button"
      ref={ref}
      variant="normal"
      themeMode={themeMode}
      {...props}
    >
      {props.children}
    </AppButton>
  )
})

Button.displayName = 'Button'
export default Button
