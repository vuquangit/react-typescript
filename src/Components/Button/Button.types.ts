import {
  SpaceProps,
  LayoutProps,
  FlexBasisProps,
  ButtonStyleProps,
  BorderProps,
  TypographyProps,
} from 'styled-system'

interface ThemeButton {
  color?: string
  kind?: string
}

interface ButtonProps {
  theme?: ThemeButton
  label?: string
  size?: 'small' | 'medium' | 'large'
}

export declare type BaseButtonProps = React.ButtonHTMLAttributes<
  HTMLButtonElement
> &
  ButtonProps &
  SpaceProps &
  LayoutProps &
  FlexBasisProps &
  ButtonStyleProps &
  BorderProps &
  TypographyProps

export interface IButtonProps extends BaseButtonProps {
  children?: React.ReactNode
  props?: BaseButtonProps
  ref?: any
}
