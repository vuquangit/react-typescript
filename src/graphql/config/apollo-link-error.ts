import { onError } from '@apollo/client/link/error'
// import { logout } from './logout';

export const errorLink = onError(
  ({ graphQLErrors, networkError, response, operation }) => {
    if (graphQLErrors) {
      for (const error of graphQLErrors) {
        console.error(
          `[GraphQL error]: Message: ${error.message}, Location: ${error.locations}, Path: ${error.path}`,
          operation,
          response
        )
      }
    }

    // if (networkError?.statusCode === 401) logout()

    if (networkError) {
      console.error(`[Network error]: ${networkError}`, operation, response)
    }
  }
)
