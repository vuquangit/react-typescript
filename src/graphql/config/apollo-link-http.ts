import { createHttpLink } from '@apollo/client'
import { setContext } from '@apollo/client/link/context'
import fetch from 'cross-fetch'

const authLink = setContext((_, { headers }) => {
  // get the authentication token from local storage if it exists
  // const token = localStorage.getItem('token')

  // return the headers to the context so httpLink can read them
  return {
    headers: {
      ...headers,
      // 'Authorization': token ? `Bearer ${token}` : null,
      // 'Content-Type': 'application/json',
      // Accept: 'application/json',

      authorization: localStorage.getItem('token') || '',
      'client-name': 'ac3-todos-backend',
      'client-version': '1.0.0',
    },
  }
})

const createdHttpLink = createHttpLink({
  uri: process.env.REACT_APP_API_ENDPOINT || '',
  fetch: fetch,
})

const httpLink = authLink.concat(createdHttpLink)

export { httpLink }
