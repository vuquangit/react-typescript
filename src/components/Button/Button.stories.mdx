import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs/blocks'
import { action } from '@storybook/addon-actions'
import Button from '.'
import Space from 'stories/components/styled-system/Space.mdx'
import Styles from 'stories/styles.mdx'

<Meta title="Story/Components/Button" component={Button} />

<Styles />

# Button

With `MDX` we can define a story for `Button` right in the middle of our
Markdown documentation.

export const Template = (args) => <Button {...args} />

<Canvas>
  <Story
    name="Normal"
    args={{ variant: 'normal' }}
    onClick={action('Medium button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Large"
    args={{ variant: 'large' }}
    onClick={action('Large button clicked')}
  >
    {Template.bind({})}
  </Story>
  <Story
    name="Disable"
    args={{ disabled: true }}
    onClick={action('Disable button clicked')}
  >
    {Template.bind({})}
  </Story>
</Canvas>

<div class="subheading">Props</div>

<Space />
