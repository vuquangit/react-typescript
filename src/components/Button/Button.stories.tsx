import React from 'react'
// also exported from '@storybook/react' if you can deal with breaking changes in 6.1
import { Story, Meta } from '@storybook/react/types-6-0'
import Wrapper from '@/test/supports/AppWrapper'

// import {
//   Title,
//   Subtitle,
//   Description,
//   Primary,
//   ArgsTable,
//   Stories,
//   PRIMARY_STORY,
// } from '@storybook/addon-docs/blocks'

import { Button } from '.'
import { IButtonProps } from './Button.types'

export default {
  title: 'Story/Components/Button',
  component: Button,
  argTypes: {
    label: {
      control: {
        type: 'text',
      },
    },
    variant: {
      control: {
        type: 'select',
        options: ['normal', 'large'],
      },
    },
    disabled: {
      control: {
        type: 'boolean',
      },
    },
  },
} as Meta

const Template: Story<IButtonProps> = (args) => (
  <Wrapper themeMode="light">
    <Button {...args}>Button story</Button>
  </Wrapper>
)

export const Normal = Template.bind({})
Normal.args = {
  label: 'Normal button',
  variant: 'normal',
}

export const Large = Template.bind({})
Large.args = {
  label: 'Large button',
  variant: 'large',
}

export const Disable = Template.bind({})
Disable.args = {
  disabled: true,
}
