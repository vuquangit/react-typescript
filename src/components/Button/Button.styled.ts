import styled, { css } from 'styled-components'
import { prop, ifProp, switchProp } from 'styled-tools'

import {
  space,
  layout,
  flexbox,
  buttonStyle,
  border,
  typography,
  variant,
  color,
} from 'styled-system'
import { BaseButtonProps } from './Button.types'

export const AppButton = styled('button')<BaseButtonProps>`
  ${color};
  ${space};
  ${layout};
  ${flexbox};
  ${buttonStyle};
  ${border};
  ${typography};

  cursor: ${prop('cursor', 'default')};
  border: 0;

  ${switchProp('themeMode', {
    light: css`
      background-color: #000;
      color: #fff;
    `,
    dark: css`
      background-color: #fff;
      color: #000;
      border: 1px solid #fff;
    `,
  })}

  ${ifProp(
    'disabled',
    css`
      background-color: ${prop('theme.colors.gray', '#999')};
      border-color: ${prop('theme.colors.gray', '#999')};
      pointer-events: none;
    `
  )}

  ${variant({
    variants: {
      normal: {
        // p: '8px',
        fontSize: '16px',
      },
      large: {
        // p: '12px',
        fontSize: '24px',
      },
    },
  })}
`
