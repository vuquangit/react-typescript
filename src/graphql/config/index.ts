import { ApolloClient, NormalizedCacheObject, ApolloLink } from '@apollo/client'

import { httpLink } from './apollo-link-http'
import { errorLink } from './apollo-link-error'
import { localCache, initLocalCache } from './apollo-local-cache'
import { localResolvers } from './apollo-resolvers'
import { restLink } from './apollo-link-rest'

export const apolloClient: ApolloClient<NormalizedCacheObject> = new ApolloClient(
  {
    link: ApolloLink.from([errorLink, restLink, httpLink]),
    // Note: httpLink is terminating so must be last, while retry & error wrap
    // the links to their right. State & context links should happen before (to
    // the left of) restLink.
    connectToDevTools: process.env.NODE_ENV !== 'production',
    cache: localCache,
    assumeImmutableResults: true,
    resolvers: localResolvers,
  }
)

initLocalCache()
