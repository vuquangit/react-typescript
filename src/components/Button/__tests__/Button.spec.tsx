import React from 'react'
import { render } from '@testing-library/react'
import AppWrapper from '@/test/supports/AppWrapper'

import Button from '..'
import { applyTheme } from '@/graphql/config/apollo-local-cache'

describe('App Button', () => {
  it('Test Button with light mode', async () => {
    applyTheme('light')

    const { container, getByTestId } = render(
      <AppWrapper themeMode="light">
        <Button size="large">Primary</Button>
      </AppWrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'background-color': '#000',
      color: '#fff',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })

  it('Test Button with dark mode', async () => {
    applyTheme('dark')

    const { container, getByTestId } = render(
      <AppWrapper themeMode="dark">
        <Button size="large">Primary</Button>
      </AppWrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'background-color': '#fff',
      color: '#000',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })

  it('Test Button disabled', async () => {
    applyTheme('light')

    const { container, getByTestId } = render(
      <AppWrapper themeMode="light">
        <Button disabled>Primary</Button>
      </AppWrapper>
    )

    expect(container).toMatchSnapshot()

    const expectStyle: { [key: string]: string } = {
      'pointer-events': 'none',
    }

    for (const style in expectStyle) {
      if (expectStyle[style]) {
        expect(getByTestId('app-button')).toHaveStyleRule(
          style,
          expectStyle[style]
        )
      }
    }
  })
})
