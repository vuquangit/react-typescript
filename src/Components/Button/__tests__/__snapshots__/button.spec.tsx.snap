// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button disable btn 1`] = `
.c0 {
  width: medium;
  height: medium;
  cursor: default;
  color: #fff;
  background-color: #000;
  font-size: 14px;
  background-color: #999;
  border-color: #999;
  pointer-events: none;
}

<Button
  disabled={true}
>
  <Component
    theme={
      Object {
        "palette": Object {
          "primary": Object {
            "contrastText": "#FFF",
            "dark": "#cf2557",
            "light": "#fc688e",
            "main": "#F62F5E",
          },
          "secondary": Array [
            "#c2185b",
            "#e91e63",
            "#f06292",
            "#f8bbd0",
          ],
        },
      }
    }
  >
    <styled.button
      disabled={true}
      size="medium"
      theme={
        Object {
          "kind": "dark",
        }
      }
      variant="normal"
    >
      <button
        className="c0"
        disabled={true}
        size="medium"
      >
        Disable button
      </button>
    </styled.button>
  </Component>
</Button>
`;

exports[`Button primary large btn 1`] = `
.c0 {
  width: large;
  height: large;
  cursor: default;
  color: #fff;
  background-color: #000;
  font-size: 20px;
}

<Button
  size="large"
>
  <Component
    theme={
      Object {
        "palette": Object {
          "primary": Object {
            "contrastText": "#FFF",
            "dark": "#cf2557",
            "light": "#fc688e",
            "main": "#F62F5E",
          },
          "secondary": Array [
            "#c2185b",
            "#e91e63",
            "#f06292",
            "#f8bbd0",
          ],
        },
      }
    }
  >
    <styled.button
      size="large"
      theme={
        Object {
          "kind": "dark",
        }
      }
      variant="normal"
    >
      <button
        className="c0"
        size="large"
      >
        Primary
      </button>
    </styled.button>
  </Component>
</Button>
`;
